### PR TITLE
Prelu support

### PR DESCRIPTION
**Pull request by PhilippKopp**
Hi Tobias, I am trying to add PReLU layer support to fdeep. With the modifications so far the networks pass the test when loading the json file in most cases. The only case where it does not pass the test is when the PReLU layer comes after a conv layer, which happens to be quite often and also in the network I want to use this for. I don't really understand how the layer before influences a layer coming behind it.
I know that what I have so far can be more optimized and is far from perfect, but the first step would be to get it running and get the same results as in keras. Do you have any ideas what might be the problem?

### CODE REVIEW

**Comment by Dobiasd** [comment]
Hi Philipp,

nice work. In general this looks quite good.

Of course you are right, for the PReLU layer it should not matter if it comes behind a conv layer or not. Perhaps the input shape of the PReLU layer was 3-dimensional in your tests when you used it after the conv layer but 2/1-dimensional in your other tests? My wild guess is that the order of the weights is not correct (channel last vs. channel first) and you have to use something like
```python
weights_flat = np.swapaxes(weights[0], 0, 2).flatten()
```
in your `show_prelu_layer` function.

In any case, to make testing easier, I suggest you add the following three lines to `get_test_model_small` in `generate_test_models.py`:
```python
    outputs.append(keras.layers.PReLU()(inputs[0]))
    outputs.append(keras.layers.PReLU()(inputs[1]))
    outputs.append(keras.layers.PReLU()(inputs[2]))
```

Then the CI will help us run the tests, and we will get feedback here directly attached to this pull request. And the tests would also include input shapes with different numbers of dimensions, based on the structure of `inputs`.

For the CI however you should also first fix the compiler warnings, since they result in errors with our settings, see: https://travis-ci.org/Dobiasd/frugally-deep/builds/389705313
(click `Raw log`) to see everything as plain text.

**Comment by PhilippKopp** [comment]
You really have an impressive testing pipeline here!
I think I solved all the compiler warnings and the CI ran further than before. I also added the PReLUs to `get_test_model_small` . If I understand the error message correctly it now crashes in the sequential test, which I didn't touch at all (at least not deliberately). Do you understand what's going on?

I had tested the PReLU implementation on 3D input tensors directly and it passed the tests. So it seems it's not only 3D input where the bug comes into play. I also tried your suggestion and that didn't help either. Maybe you can read more out of the CI. Thanks a lot!

**Comment by Dobiasd** [comment]
Your changes look good.

The error in CI is really strange.
```
/home/travis/build/Dobiasd/frugally-deep/test/test_model_sequential_test.cpp:11: ERROR: test case THREW exception: [json.exception.type_error.305] cannot use operator[] with array
```
I'll investigate.

**Comment by Dobiasd** [comment]
Just re-ran the CI pipeline of the master branch, and it [also failed](https://travis-ci.org/Dobiasd/frugally-deep/builds/386070402). When I originally made the last commit 9 days ago, it was working.

The new error seems to stem from the latest Keras version. With Keras `2.1.6` it is fine, with Keras `2.2.0` it is not. I will work on this, but to unblock further development I set [`2.1.6` in the Travis script](https://github.com/Dobiasd/frugally-deep/commit/09b8b03f49007a7a35437fdbb939488521d912f8), and not [it's OK again](https://travis-ci.org/Dobiasd/frugally-deep/builds/390131037).

I recommend you merge this commit into your branch (should not create and conflict). Then we can can again test the actual feature. :)

You can also run the unit tests locally if you like:
```bash
# from inside your frugally-deep directory:
mkdir -p build
cd build
cmake -DFDEEP_BUILD_UNITTEST=ON ..
make unittest
```

**Comment by Dobiasd** [comment]
Nice! Currently it seems like we now have correctly working support for PReLU layers. :grin:

Can you please try to recreate the issue you had with PReLU after convolution in the small test model?
If the issue just won't re-appear with the test model, you can also send the model that produces the problem directly to me.

**Comment by PhilippKopp** [comment]
Would be nice :D
It took me a while to get the unit tests running locally. It works now and I could reproduce the errors I saw before. It doesn't pass the test anymore when applying PReLU after a convolution layer.

**Comment by Dobiasd** [comment]
Very interesting. I will checkout your PR locally and see if I can find the cause.

Would you mind telling me the problems you had in getting the tests to execute? Perhaps there is room for improving the documentation?

**Comment by PhilippKopp** [comment]
That would be awesome!! Thank you very much!

The main issues were because I am using windows. So I had to adapt the compile options https://github.com/Dobiasd/frugally-deep/blob/master/cmake/toolchain.cmake#L1 for VS and all the commands for executing the python scripts: https://github.com/Dobiasd/frugally-deep/blob/master/test/CMakeLists.txt#L8

**Comment by Dobiasd** [comment]
OK, understood. I used Windows to develop in my job for years, and although Visual C++ is a great IDE, building C++ projects with dependencies can be quite a pain. So I'd like to avoid adding build instructions for windows to the frugally-deep docs. :wink:
If you work on a library like frugally-deep, using Docker or even a Virtualbox with a simple Ubuntu 16.04 or something could perhaps make things more smooth.

---

Regarding the actual problem here: I can reproduce it locally, and currently it is still a mystery to me. I'll report back when I found out what is happening.

**Comment by Dobiasd** [comment]
Please merge the latest commits from the master branch and then use the following in `show_prelu_layer`:

```python
    if len(weights[0].shape) == 3:
        weights_flat = np.swapaxes(np.swapaxes(weights[0], 2, 1), 1, 0).flatten()
    elif len(weights[0].shape) == 2:
        weights_flat = np.swapaxes(weights[0], 1, 0).flatten()
    elif len(weights[0].shape) == 1:
        weights_flat = weights[0].flatten()
    else:
        assert False
```

**Comment by Dobiasd** [comment]
Yeah, looks good. What do you think? Ready to merge? :slightly_smiling_face:

**Comment by PhilippKopp** [comment]
Yeah really nice! Thank you so much :) 
Also just tested my Networks and they also pass the tests!

Are you okay with the implementation of the layer itself? https://github.com/PhilippKopp/frugally-deep/blob/prelu_support/include/fdeep/layers/prelu_layer.hpp#L24

Currently I save the alphas in a vector as class member, "convert" them to a tensor and then access them through it. We could also create the tensor only once in the constructor or access the correct alpha from the vector directly as in the cout comment. What would you prefer? If you're fine with how it is, I'll remove the comments and we're ready to merge :)

**Comment by Dobiasd** [comment]
In case you like to know how I found out about the order of axis : Basically just [trying out different `swapaxes` combinations](https://i.imgur.com/KKgg3dd.png) until it worked. ;)
I have to admit I had similar problems in the past with the parameters of the other layer types.

I'm sorry for not yet having a better process to generate exports for new layers with proper documentation. Thanks for continuing nonetheless. :)

Regarding your question: Currently we are making a full in-memory copy of all values in `alpha_` with every call to `apply_impl`. For performance it would be better to avoid this. But we can not store it as a `tensor` as a member of the `prelu_layer`, since we need to know the tensor shape, and we only get to know it during prediction. Storing the shape with the values and loading it would not match the storage format of the other layer types and could hinder future changes in the file format.

But there is a simply solution:
Have an `fplus::shared_ref<std::vector<fdeep::float_type>>` (`fdeep::float_type` instead of float because it can be changed) aka. `fdeep::shared_float_vec` as a member of `prelu_layer`, that is created with `fplus::make_shared_ref` in the member initialization list the constructor of `prelu_layer`. When we then use it to create a `tensor3` on the fly at the start of `apply_impl` the values will not be copied, but the tensor will just reuse the underlying memory. A mismatch of the target shape and number of values in the vector would cause the constructor of `tensor3` to raise an exception of course.

If you implement this and clean-up the debug comments, I guess we can pour the champagne. :D

**Comment by PhilippKopp** [comment]
Hahaha :D I was about to ask :) You must still have some good heuristics about what axes to choose. With a double `swapaxes` there are quite a number of possible combinations :) 

Oh yeah, didn't think about the tensor shape. Implemented your solution now! 
I thinks it's time for the champagne!! Thanks a lot for your help :)

**Comment by Dobiasd** [comment]
Just went through the new code once more. Good work. Thanks again for the contribution. :)

Some things I adjusted:
- [disallow usage of shared_axes in PReLU](https://github.com/Dobiasd/frugally-deep/commit/129e4b3aa5718fa319f7f5561ed300eecd9f29a2), since we don't support it without this check the C++ application would raise an exception during prediction. Failing earlier is better. :)
- [added PReLU to the list of supported layer types in the `README` file](https://github.com/Dobiasd/frugally-deep/commit/09641a990e13ee80a5a6c777355823470ecc29dc)
- [adjusted indentation style to the rest of the lib](https://github.com/Dobiasd/frugally-deep/commit/f8c138a93c1f67fa4d4b5bef6b58b2f379172e1b)
- [Changed the order of the for loops in `prelu_layer.hpp` from `x, y, z` to `z, y, x`](https://github.com/Dobiasd/frugally-deep/commit/252e0e28970ecd2482ac4cb9ee550d760e9ddeeb). Due to the [order of the values of a `tensor3` in memory](https://github.com/Dobiasd/frugally-deep/blob/f3354de109eed6292e5e3ce48ae4c7597c73b9a8/include/fdeep/tensor3.hpp#L94-L97) this is better to avoid CPU-cache misses and thus improve performance.

**Comment by PhilippKopp** [comment]
I have to thank you for your help! There's not much of my initial code left :D You did most of the work :) I did learn a lot though, Thanks!
